test/Generators.hs

```diff
@@ -15,17 +15,48 @@ import qualified Hedgehog.Range as Range
 
 -- DISCREPANCY: 'jq' allows both 000 in input and its own number literals,
 -- but JSON spec does not. We'll stay closest to 'jq' by doing this, too.
+--
 
--- First, an unsized Gen of JSON numbers.
+-- | Helper Generators
+signGen :: Gen Text
+signGen = Gen.element ["+", "-", ""]
 
+signGen' :: Gen Text
+signGen' = Gen.element ["-", ""]
+
+integerGen :: Gen Text
+integerGen = Text.pack . show <$>
+      (Gen.integral $ Range.exponential 1 (10^309))
+
+expGen :: Gen Text
+expGen = Text.pack . show <$>
+  (Gen.integral (Range.linearFrom 0 (-10^16) (10^16)))
+
+decBefore :: Gen Text
+decBefore = Gen.constant "." <> integerGen
+
+decAfter :: Gen Text
+decAfter = integerGen <> Gen.constant "."
+
+-- | Number type generators
 numberGen :: Gen Text
-numberGen = sign <> integer <> commaPart
+numberGen = signGen' <> integerGen
+
+scientificNumberGen :: Gen Text
+scientificNumberGen = base <> e <> signGen <> expGen
   where
-    sign = Gen.element ["", "-"]
-    integer = Text.singleton <$> Gen.element ['0'..'9']
-    commaPart = Gen.element [ "..." ]
+    base = Text.singleton <$> Gen.element ['0'..'9']
+    e = Gen.element ["e", "E"]
+
+fractionalGen :: Gen Text
+fractionalGen = integerGen <> Gen.constant "." <> integerGen
 
--- Second, a sized one.
+optFractionalGen :: Gen Text
+optFractionalGen = signGen' <> Gen.choice [decBefore, decAfter]
```

Comment:
These cover `<x>.`, `-<x>.`, `.<y>` and `-.<y>`, check.

Then what about:
- `<x>.e...`
- `.<y>e...`
- `-<x>.e...`
- `-.<y>e...`

I understand that this generator is called `optFractionalGen`, and is one of the four generators that are referred to below, so maybe it makes sense to address these in `scientificNumberGen`; then again, they're scientific numbers with optional fractions.

I would probably have divided the sub-parsers a little differently, so that once you've made a parser that generates all combinations without scientific notation, you tack `<> Gen.choice [ "", ePart ]` onto it.

Reply:
Then the structure of the generator will somewhat resemble the JSON number CFG in #5.